### ISSUE TITLE
[RFC] Qwen-Image DMD2 adapter and MVP validation

### ISSUE THREAD

**Issue by NancyFyong** [verl-project/verl-omni#543]
## Feature request

Complete the **Qwen-Image DMD2 distribution-only MVP** in #519 using the loss/config work in #520 and the concrete engine/trainer integration in #535.

Following [the review](https://github.com/verl-project/verl-omni/pull/545#issuecomment-5609385564), replace the combined DMD/DMD2 scope with one real model/algorithm path. This can land with the runtime PR, not as a third framework layer. Reuse [the Qwen prototype](https://github.com/NancyFyong/verl-omni/pull/9), but do not treat its old results as refactor acceptance.

## Motivation

The MVP is selected by **`actor_rollout_ref.model.algorithm=dmd2`** and trains from prompts. It uses a distribution-matching student objective, a generated-sample fake-score denoising objective, separate optimizers and a two-timescale update schedule.

It is **not original DMD**: there is no paired `reference_noise`/teacher-target regression, LPIPS dependency or regression data loader. Reserve `dmd` for the future original-DMD implementation; do not stack `dmd` and `dmd2` registrations on one adapter and claim both algorithms are supported. DMD2 GAN, consistency-renoise ablations, causal Wan and ODE training stages are not required for this MVP.

Training is offline in the framework sense: prompts come from the dataset and differentiable samples are generated inside the training worker. There is no RL reward/PickScore training objective, detached vLLM training rollout, PPO log-probability batch or advantage estimator.

## Qwen implementation boundary

Register the concrete `(QwenImagePipeline, dmd2)` training adapter through the existing registry and pipeline imports. The target is base Qwen-Image T2I; edit checkpoints and other pipeline class names are not implicitly supported.

Keep conditioning, packing and scheduler/forward adaptation in `verl_omni/pipelines/qwen_image_distillation/diffusers_training_adapter.py`; the engine owns sampling/optimization. Reuse Qwen helpers, not a forked forward stack or trainer architecture branches. Add only methods the engine uses, not a mandatory computer/capability/recipe hierarchy or new abstract methods for unrelated pipelines.

### Qwen flow and latent geometry

Read VAE geometry/normalization from the checkpoint. Work with normalized image latents, pack/unpack using Qwen helpers, and validate the declared image resolution and packing alignment. Do not guess a layout from arbitrary tensor dimensions or hardcode model size.

```text
x_sigma = (1 - sigma) * x0 + sigma * epsilon
v_target = epsilon - x0
x0_prediction = x_sigma - sigma * velocity_prediction
```

The transformer uses packed image tokens and normalized model timesteps in `[0,1]`. Preserve fp32 latent/sigma/transition/score arithmetic around mixed-precision forwards. Check packing round trips and model-output dimensions explicitly; tensor transport must not confuse conditioning latents with the denoised sample.

### Conditioning

Reuse the existing tested Qwen conditioning code:

- Plain strings/single text-only user messages use `QwenImagePipeline.prompt_template_encode` and its real prefix-removal contract, not generic chat formatting that empties short prompts.
- Unsupported chat structures fail clearly or require precomputed conditioning, never silent flattening.
- IDs require masks; `[B,L,D]` embeddings require compatible masks/batches and length limits. Frozen outputs are detached; materialize all-ones masks when existing SP padding needs them.
- Negative-condition caching handles batch size correctly. Teacher scoring needs it; the conditional-only fake stage need not encode it.

Use existing prompt-parquet/collation and accumulation, with tested same-resolution physical batches. Do not add paired-regression columns or unused generic containers.

### Student rollout and scoring

For four-step inference, the proposed fixed rational-shift-3 grid is:

```text
sigmas = [1.0, 0.9, 0.75, 0.5, 0.0]
x_next = x + (sigma_next - sigma) * velocity
```

Use one shared schedule builder for training and any retained inference adapter. This fixed-grid choice deliberately differs from the native resolution-dependent Qwen schedule; do not apply both shifts or claim parity with a different native schedule.

Sample a training exit index, run preceding Euler steps under no-grad and retain the exit graph. Preserve evaluation/dropout behavior through checkpoint recomputation. Fake-stage student sampling is entirely no-grad. Validate the four-step baseline; fewer steps do not imply original-DMD support.

Broadcast collective-dependent exits/branches across FSDP/model-parallel ranks. Tensor noise may differ; forward counts and gradient exits may not. Checkpoint initial-noise, exit, score-sigma and score-noise RNG streams.

The proposed baseline discrete score sampler draws an integer from `[0,T)`, normalizes by the scheduler's actual training-step count `T`, applies the chosen shift once and clamps to the configured sigma bounds. The Qwen baseline uses `T=1000`, shift `3.0` and bounds `[0.02,0.98]`. Any retained continuous-uniform option samples directly inside the bounds without discrete remapping or endpoint-clipping a full-range draw. Record the selected sampling mode and values; these are explicit Qwen adaptation choices, not a claim to reproduce every reference configuration.

Teacher and fake score use the same re-noised sample for #520's student objective. Apply teacher CFG as `negative + scale*(positive-negative)` and normalize **in packed denoiser velocity space, before unpacking/x0 conversion**. Fake score and student are conditional-only. Do not inherit the teacher scale for fake scoring, mix prediction parameterizations or silently substitute another Qwen CFG variant.

## MVP example and runtime integration

Use top-level `dmd:` configuration, `algorithm.trainer_type=distribution_matching` and `algorithm.sample_source=offline`; leave OPD's `distillation:` fields alone. The new settings are proposed until #520/#535 land and are not a command supported by old prototype heads.

The initial documented example should make these choices explicit:

| Setting | Baseline / requirement |
| --- | --- |
| Model | Pinned base Qwen-Image revision, T2I |
| Resolution / sampling | 1024 x 1024, four-step deterministic inference, fixed shift 3 |
| Student/fake LoRA | Rank/alpha 32, targets `to_q`, `to_k`, `to_v`, `to_out.0` |
| Student/fake learning rate | `1e-4` / `2e-5`, independently configured optimizers |
| Fake attempts per student attempt | `K=2`; report actual successful counts and skips |
| Teacher guidance | Scale 4, `layer_norm`, explicit negative prompt such as a single space |
| Student/fake guidance | Conditional-only |
| Warmup, EMA and seed | Explicitly record their values and successful-update/RNG units; do not silently inherit an unrelated recipe |
| Default inference artifact | Student LoRA with matching base/config/schedule metadata |

These reuse the pinned LightX2V example's useful flow-model choices, but a filename or `method: dmd` label in another project is not sufficient to classify a run as original DMD. Claims of exact reproduction require checking the actual objectives, sampler and all run settings.

Use the shared-base named-adapter baseline and #535's independent optimization/skip contract. Reuse microbatch/DP machinery. Replay exits for batching comparisons: physical batches share an exit, accumulation need not. Scalar CFG also changes with batch composition.

Preserve tested templates, batch-safe conditioning, synchronized exits, normalization, gradient isolation, RNG and adapter-export fixes; do not revert them for a smaller layout.

## Checkpoint and inference acceptance

Restore both optimizers/schedulers, student/fake/EMA, data/RNG and #535's counters; file presence is not resume validation.

Export **student** with its own PEFT config, base revision and exact schedule; reload into a fresh model and decode images. Declare the base-checkpoint dependency. EMA is an explicit alternative, not a mandatory second export.

Reuse existing deterministic inference code. Any vLLM parity claim needs the pinned runtime, matching initial latents, embeddings/masks, schedule/guidance and padded-text handling; scheduler-only tests are insufficient. Validation-replica orchestration, batched throughput and new serving/transport are deferred. The MVP gate is training, complete resume, student reload and generation.

## Validation plan

- [ ] CPU loss/config/adapter tests resolve `dmd2` without advertising original `dmd`; malformed or out-of-scope configuration fails clearly.
- [ ] Realistic short prompts preserve nonempty conditioning; precomputed/token-ID inputs and same-resolution batches retain valid masks and batch dimensions.
- [ ] Packing, flow conversion, guidance, score sampling and fixed-grid transitions match explicit numerical fixtures and precision boundaries.
- [ ] Student gradients survive teacher/fake scoring; fake training has no path into student parameters. EMA changes only after an eligible successful student update.
- [ ] Real multi-rank FSDP1/FSDP2 runs execute repeated cycles with synchronized exits, tested microbatching and no collective-order divergence.
- [ ] Explicit numerical skips preserve scheduler/EMA/success counts and finite termination; unexpected failures require complete-checkpoint recovery.
- [ ] A real Qwen multi-GPU smoke completes multiple successful student/fake updates, saves and resumes, and reloads the exported student artifact.
- [ ] A controlled new-versus-prototype comparison starts from the same weights, data and RNG and checks losses, gradients/updates and resume behavior with stated tolerances.
- [ ] Existing OPD, DPO/NFT and policy-gradient regressions plus generated-config, lint/type and sanity checks pass on recorded revisions.

Record component timings and peak memory using existing profiling, retaining every fake attempt and correct aggregation. Do not mix nested host timings, CPU means and GPU wall time into an invented speedup.

For an effectiveness report, use held-out prompts/seeds and matched initial noise. Include base long-step, base few-step with matched student guidance, and distilled few-step samples so weight effects are not confused with schedule/CFG changes. Label exploratory sample sizes and metric limitations. Finite loss, a tiny-model smoke or a few images alone does not establish quality parity, improvement or isolated latency.

## Your contribution

Coordinate this real model consumer with #535 instead of treating it as a later justification for an already-fixed generic framework. Original-DMD paired regression and its optional perceptual dependency remain separate future work; so do GAN, CausVid and Self-Forcing. Existing branches and evaluation artifacts are preserved, not retroactively relabeled as MVP evidence.

The checklist applies to the revised implementation and starts unverified. Record actual commands, code/model/dependency revisions, results and limitations in the eventual code PRs, with the required human review and contribution disclosures.

References: [parent #519](https://github.com/verl-project/verl-omni/issues/519), [loss/config #520](https://github.com/verl-project/verl-omni/issues/520), [runtime #535](https://github.com/verl-project/verl-omni/issues/535), [reviewer proposal](https://github.com/verl-project/verl-omni/pull/545#issuecomment-5609385564), and [pinned Qwen flow-model example](https://github.com/ModelTC/LightX2V/blob/2a603235e07ed281879ae90c5624abc05a14c35b/lightx2v_train/configs/train/dmd/qwen_image_dmd_lora.yaml).
